openvdb_houdini/houdini/PointUtils.cc

```diff
@@ -926,110 +926,122 @@ convertHoudiniToPointDataGrid(const GU_Detail& ptGeo,
+    assert(numVDBPoints <= numHoudiniPoints);
 
-    // Append (empty) groups to tree
+    if (numVDBPoints < numHoudiniPoints) {
+        warnings("Points contain NAN positional values. These points will not be converted.");
+    }
 
-    std::vector<Name> groupNames;
-    groupNames.reserve(elementGroups.entries());
+    if (!tree.cbeginLeaf())  return pointDataGrid;
 
-    for (auto it = elementGroups.beginTraverse(), itEnd = elementGroups.endTraverse();
-        it != itEnd; ++it)
-    {
-        groupNames.push_back((*it)->getName().toStdString());
-    }
+    // store point group information
 
-    appendGroups(tree, groupNames);
+    const GA_ElementGroupTable& elementGroups = ptGeo.getElementGroupTable(GA_ATTRIB_POINT);
+    const size_t numGroups = elementGroups.entries(); // including internal groups
 
-    // Set group membership in tree
+    if (numGroups > 0) {
 
-    const int64_t numPoints = ptGeo.getNumPoints();
-    std::vector<short> inGroup(numPoints, short(0));
+        // Append (empty) groups to tree
 
-    for (auto it = elementGroups.beginTraverse(), itEnd = elementGroups.endTraverse();
-        it != itEnd; ++it)
-    {
-        // insert group offsets
-
-        GA_Offset start, end;
-        GA_Range range(**it);
-        for (GA_Iterator rangeIt = range.begin(); rangeIt.blockAdvance(start, end); ) {
-            end = std::min(end, GA_Offset(numPoints));
-            for (GA_Offset off = start; off < end; ++off) {
-                UT_ASSERT(off < GA_Offset(numPoints));
-                inGroup[off] = short(1);
-            }
+        std::vector<Name> groupNames;
+        groupNames.reserve(numGroups);
+
+        for (auto it = elementGroups.beginTraverse(), itEnd = elementGroups.endTraverse();
+            it != itEnd; ++it) {
+            groupNames.emplace_back((*it)->getName().toStdString());
         }
 
-        const Name groupName = (*it)->getName().toStdString();
-        setGroup(tree, indexTree, inGroup, groupName);
+        appendGroups(tree, groupNames);
+
+        std::vector<short> inGroup(numHoudiniPoints, short(0));
 
-        std::fill(inGroup.begin(), inGroup.end(), short(0));
+        // Set group membership in tree
+        // @todo parallelize group membership construction
+
+        for (auto it = elementGroups.beginTraverse(), itEnd = elementGroups.endTraverse();
+            it != itEnd; ++it) {
+
+            GA_Offset start, end;
+            GA_Range range(**it);
+            for (GA_Iterator rangeIt = range.begin(); rangeIt.blockAdvance(start, end); ) {
+                end = std::min(end, numHoudiniPoints);
+                for (GA_Offset off = start; off < end; ++off) {
+                    UT_ASSERT(off < GA_Offset(numHoudiniPoints));
+                    inGroup[off] = short(1);
+                }
+            }
+
+            const Name groupName = (*it)->getName().toStdString();
+            setGroup(tree, indexTree, inGroup, groupName);
+
+            // reset groups to 0
```

Comment:
Running it through tbb for progressively larger data sets should be heaps faster - you're referring to the size of the vector right? Here's a very basic test using -O3 https://s22.postimg.cc/wkaix1gcx/fillvstbb.png (orange is stl, x is elements, y is time)

Reply:
My point is that there are some low-level optimizations you miss out on by directly indexing instead of using std::fill().

Try this as an alternative:

```
tbb::parallel_for(tbb::blocked_range<size_t>(0, inGroup.size()/1024),
    [&](const tbb::blocked_range<size_t>& range) {
        for (size_t n = range.begin(), N = range.end(); n != N; ++n) {
            std::fill_n(inGroup.begin() + n*1024, 1024, short(0));
        }
});
```

You obviously have to expand the inGroup array to be a multiple of 1024 to avoid having to do any range checking.

Reply:
Ah I see, as long as the inGroup vector is created at a multiple of 1024 I think you're right, otherwise the possible reallocation will quickly engulf the overall performance. Simple range checking should also be faster, but I'll change this to the first method (though this all probably a tad overkill).